### PR TITLE
feat(colors): directive to use any color from any palette everywhere

### LINKED ISSUE #1269: colors and hues should be available everywhere.
I created [this sample on plunker](http://plnkr.co/edit/M22sy0AI7Bax3Cop3OmI?p=preview) to demonstrate a few issues I'm having with forms. 

If you look at like 24 of index.html in my sample, you see there is a heading there. This heading is not visible in the page, because I don't know how to set the background color according to the current selected theme/colorpallete. I added an md-hue-1 to the class-list as well as an attribute, but it is not giving the result I need.

This is also mentioned in #537, but I still could not find a solution to it.

### PR DESCRIPTION
**Pull request by EladBezalel**
- added directive and documentation
- added tests
- added smart watch option, when the directive recognize at least one of it's values needs interpolation it starts to watching it

fixes #1269

### CODE REVIEW

File: src/components/colors/colors.js [hunk 1]
```diff
@@ -0,0 +1,88 @@
...
+ * </hljs>
+ *
+ */
+
+function mdColorsDirective($mdColorPalette, $mdTheming) {
+  return {
+    restrict: 'A',
+    link: function (scope, elem, attrs) {
+      scope.$watch(attrs.mdColors, function (val) {
+        Object.keys(val).forEach(function (key) {
```
Comment (crisbeto) [inline src/components/colors/colors.js]: You could spare calling `Object.keys` and iterating it by using `angular.forEach(function(value, key){ ... });`

File: src/components/colors/colors.js [hunk 2]
```diff
@@ -0,0 +1,88 @@
...
+          if (Object.keys($mdColorPalette).indexOf(split[1]) !== -1) {
+            palette += '-' + split.splice(1, 1);
+          }
+          else {
+            // Two word palette can be also written in camelCase, forming camelCase to dash-case
+            palette = palette.replace(/([a-z])([A-Z])/g, '$1-$2').toLowerCase();
+          }
+
+          // If the palette is not in the palette list it's one of primary/accent/warn/background
+          if (Object.keys($mdColorPalette).indexOf(palette) === -1) {
```
Comment (crisbeto) [inline src/components/colors/colors.js]: You're calling `Object.keys` on `$mdColorPalette` twice in the same function, you could call it once and save it in a variable.

File: src/components/colors/colors.js [hunk 3]
```diff
@@ -0,0 +1,88 @@
...
+            var scheme = $mdTheming.THEMES[theme].colors[palette];
+
+            if (!scheme) {
+              return console.error('mdColors: couldn\'t find \'' + palette + '\' in the palettes.');
```
Comment (crisbeto) [inline src/components/colors/colors.js]: Not really a big deal, but you could use `$log.error` in here.

File: src/components/colors/colors.js [hunk 4]
```diff
@@ -0,0 +1,88 @@
...
+ *   </div>
+ * </hljs>
+ *
+ */
+
+function mdColorsDirective($mdColorPalette, $mdTheming) {
+  return {
+    restrict: 'A',
+    link: function (scope, elem, attrs) {
+      scope.$watch(attrs.mdColors, function (val) {
```
Comment (DevVersion) [inline src/components/colors/colors.js]: Just a question, why are you not using `$observe`, are there problems with the interpolating?

File: src/components/colors/colors.js (hunk 1 above)
Comment (EladBezalel) [inline src/components/colors/colors.js]: I encountered today something even more beautiful!
``` js
for (key in val) {
 //...
}
```

File: src/components/colors/colors.js (hunk 3 above)
Comment (ThomasBurleson) [inline src/components/colors/colors.js]: We should **never** use `console.error()` or the `console.<fn>` methods... always use `$log.<fn>`.

File: src/components/colors/colors.js (hunk 4 above)
Comment (ThomasBurleson) [inline src/components/colors/colors.js]: Also `.$watch( ... )` should be consider an performance **expensive** solution.

File: src/components/colors/demoBasicUsage/script.js [hunk 5]
```diff
@@ -0,0 +1,19 @@
+(function () {
```
Comment (crisbeto) [inline src/components/colors/demoBasicUsage/script.js]: I believe the IIFE isn't necessary in here, you can inline the `DemoCtrl`.

**Comment by ErinCoughlan** [comment]
@gmoothart Is anything missing here that we could have used for tabs?

**Comment by ThomasBurleson** [comment]
@EladBezalel - The demo needs some text to describe the demo, the intent, and the power of mdColors.

**Comment by ThomasBurleson** [comment]
In a right-aligned column (in the color boxe), please show the hex color values. e.g. `#2196F3`.

screen shot 2016-04-03 at 6 00 27 pm

**Comment by ThomasBurleson** [comment]
Please refactor the `colors.spec.js` to reduce the non-DRY code.

**Comment by topherfangio** [comment]
@EladBezalel If we introduce this API, I can see developers wanting to use it in many places/components. Can you test the following components to make sure that it is compatible:

- Toast (particularly, can you easily change the background color)
- Icon
- Divider
- Subheader
- and Tooltip

If it doesn't work out of the box, we may need to update some of the components to make this easier.

**Comment by ThomasBurleson** [comment]
@EladBezalel - see refactored [`colors.spec.js`](https://gist.github.com/ThomasBurleson/979777137ddebe755c8ba0a7ee71cacd)

**Comment by EladBezalel** [comment]
- [ ] Toast (particularly, can you easily change the background color)
- [x] Icon
- [x] Divider
- [x] Subheader
- [ ] Tooltip

toast and tooltip have elements containing the styles

File: src/core/services/theming/theming.js [hunk 6]
```diff
@@ -61,8 +61,9 @@ var GENERATED = { };
 
 // In memory storage of defined themes and color palettes (both loaded by CSS, and user specified)
 var PALETTES;
-var THEMES;
 
+// Text Colors on light and dakr backgrounds
```
Comment (EladBezalel) [inline src/core/services/theming/theming.js]: Dark

**Pull request closed without merging**